### PR TITLE
Remove the link and dispatcher dir after last program is deleted

### LINKED ISSUE #141: Bug: bpfd crashes when you unload and then reload the same program
I am hitting a scenario on main that bpfd panics when reloading a previously unloaded program.

Steps to recreate:
1. Install with `sudo ./scripts/setup.sh install`
2. start bpfd w/ `sudo ./target/debug/bpfd`
3. Load a program `sudo ./target/debug/bpfctl load --iface eth0 --priority 100 --section-name xdp --program-type xdp ../xdp-tutorial/basic01-xdp-pass/xdp_pass_kern.o`
4. Unload the program `sudo ./target/debug/bpfctl unload --iface eth0 421db4eb-9c6c-42b2-8f8b-5fb9bd21be92`
5. Load the same program again `sudo ./target/debug/bpfctl load --iface eth0 --priority 100 --section-name xdp --program-type xdp ../xdp-tutorial/basic01-xdp-pass/xdp_pass_kern.o`
6. Observe bpfd panic
7. Try and restart bpfd again and it continues to panic (unless the host restarts).

Workaround - Rebooting the host will allow for bpfd to start again.

Once we resolve this, I will add it to IT.

- bpfctl output

```
### Load the xdp pass program ###
[fedora@ip-172-31-93-116 bpfd]$ sudo ./target/debug/bpfctl load --iface eth0 --priority 100 --section-name xdp --program-type xdp ../xdp-tutorial/basic01-xdp-pass/xdp_pass_kern.o
421db4eb-9c6c-42b2-8f8b-5fb9bd21be92
[fedora@ip-172-31-93-116 bpfd]$ sudo ./target/debug/bpfctl list --iface eth0
eth0
xdp_mode: skb

0: 421db4eb-9c6c-42b2-8f8b-5fb9bd21be92
	section-name: "xdp"
	priority: 100
	path: ../xdp-tutorial/basic01-xdp-pass/xdp_pass_kern.o
	proceed-on:

### Unload the program ###
[fedora@ip-172-31-93-116 bpfd]$ sudo ./target/debug/bpfctl unload --iface eth0 421db4eb-9c6c-42b2-8f8b-5fb9bd21be92

### Try and load the program again ###
[fedora@ip-172-31-93-116 bpfd]$ sudo ./target/debug/bpfctl load --iface eth0 --priority 100 --section-name xdp --program-type xdp ../xdp-tutorial/basic01-xdp-pass/xdp_pass_kern.o
Error: status: Unknown, message: "transport error", details: [], metadata: MetadataMap { headers: {} }
```

-bpfd output (bpfd exits)

```
### bpfd panics when the program is installed for the 2nd time ###
[fedora@ip-172-31-93-116 bpfd]$ sudo ./target/debug/bpfd
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: SyscallError { call: "bpf_link_create", io_error: Os { code: 16, kind: ResourceBusy, message: "Device or resource busy" } }', bpfd/src/server/bpf.rs:458:57
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

### Try and restart bpfd ###
sudo ./target/debug/bpfd
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: No such file or directory (os error 2)', bpfd/src/server/bpf.rs:158:66
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Ty!

### PR DESCRIPTION
**Pull request by nerdalert**
- This aims to resolve #141 and #146

Signed-off-by: Brent Salisbury <bsalisbu@redhat.com>

### CODE REVIEW

**Comment by nerdalert** [comment]
This patch can be validated with the following:

```
### Load some programs ###
$ sudo ./target/debug/bpfctl load --iface eth0 --priority 102 --section-name xdp --program-type xdp ../xdp-tutorial/basic01-xdp-pass/xdp_pass_kern.o
b172389e-d08b-4ee5-b781-e6df9ebe25e8
$ sudo ./target/debug/bpfctl load --iface eth0 --priority 102 --section-name xdp --program-type xdp ../xdp-tutorial/basic01-xdp-pass/xdp_pass_kern.o
082e2876-b70a-4ee1-8401-58ffa4b70392
$ sudo ./target/debug/bpfctl load --iface eth0 --priority 102 --section-name xdp --program-type xdp ../xdp-tutorial/basic01-xdp-pass/xdp_pass_kern.o
e80edf5d-a5a4-46fb-ba69-b74017c0ea4c

### View the dispatcher_{iface_id}_{rev} directory and the dispatcher_{iface_id}_link file ###
$ sudo ls -ls /var/run/bpfd/fs/dispatcher*
0 -rw-------. 1 root root 0 Sep 15 23:36 /var/run/bpfd/fs/dispatcher_2_link

/var/run/bpfd/fs/dispatcher_2_2:
total 0
0 -rw-------. 1 root root 0 Sep 15 23:36 link_082e2876-b70a-4ee1-8401-58ffa4b70392
0 -rw-------. 1 root root 0 Sep 15 23:36 link_b172389e-d08b-4ee5-b781-e6df9ebe25e8
0 -rw-------. 1 root root 0 Sep 15 23:36 link_e80edf5d-a5a4-46fb-ba69-b74017c0ea4c

### Unload the 3 programs added ###
sudo ./target/debug/bpfctl unload --iface eth0 b172389e-d08b-4ee5-b781-e6df9ebe25e8
sudo ./target/debug/bpfctl unload --iface eth0 082e2876-b70a-4ee1-8401-58ffa4b70392
sudo ./target/debug/bpfctl unload --iface eth0 e80edf5d-a5a4-46fb-ba69-b74017c0ea4c

### View the dispatcher dir and dispatcher link file are deleted ###
$ sudo ls -ls /var/run/bpfd/fs/dispatcher*
ls: cannot access '/var/run/bpfd/fs/dispatcher*': No such file or directory

### Restart bpfd ###
Now you can restart bpfd without a panic.
```

Once reviewed and merged along with the IT PR I will add this to IT. Ty!

**Comment by astoycos** [comment]
https://github.com/rust-lang/rust/issues/100219 Is blocking CI I think

File: bpfd/src/server/bpf.rs [hunk 1]
```diff
@@ -319,6 +319,10 @@ impl<'a> BpfManager<'a> {
                         BpfdError::Error("unable to delete persisted dispatcher data".to_string())
                     })?;
                 }
+                if !self.dispatchers.contains_key(&if_index) {
```
Comment (dave-tucker) [inline bpfd/src/server/bpf.rs:322]: You can include this in the `if let Some(old) = old {}` clause above.
We've already removed the key at &if_index so this `if`condition is essentially always true.

The only reason NOT to include it above would be if we wanted to cleanup pinned files even if we've run the cleanup code before, which I don't think is necessary.

**Review by dave-tucker — CHANGES_REQUESTED** [review]
Small nit but overall LGTM

File: bpfd/src/server/bpf.rs (hunk 1 above)
Comment (nerdalert) [reply bpfd/src/server/bpf.rs]: @dave-tucker done. Soon as IT merges I will get this tested as well. Ty!

**Review by dave-tucker — APPROVED** [review]
LGTM

**Pull request merged**